src/communication/v2/session.hpp

```diff
@@ -334,18 +356,22 @@ class Session final : public std::enable_shared_from_this<Session<TSession, TSes
   }
 
  private:
-  explicit Session(tcp::socket &&socket, TSessionData *data, ServerContext &server_context, tcp::endpoint endpoint,
-                   const std::chrono::seconds inactivity_timeout_sec, std::string_view service_name)
+  explicit Session(tcp::socket &&socket, TSessionContext *session_context, ServerContext &server_context,
+                   tcp::endpoint endpoint, const std::chrono::seconds inactivity_timeout_sec,
+                   std::string_view service_name)
       : socket_(CreateSocket(std::move(socket), server_context)),
         strand_{boost::asio::make_strand(GetExecutor())},
         output_stream_([this](const uint8_t *data, size_t len, bool have_more) { return Write(data, len, have_more); }),
-        session_(data, endpoint, input_buffer_.read_end(), &output_stream_),
-        data_{data},
+        session_{*session_context, endpoint, input_buffer_.read_end(), &output_stream_},
+        session_context_{session_context},
         endpoint_{endpoint},
         remote_endpoint_{GetRemoteEndpoint()},
         service_name_{service_name},
         timeout_seconds_(inactivity_timeout_sec),
         timeout_timer_(GetExecutor()) {
+#ifdef MG_ENTERPRISE
+    session_context_->Register(session_);
```

Comment:
As I understood now we have cyclic dependency between session_context and session.
Do we have plans in next release to get rid of it?

Reply:
Just to clarify a bit (the code is a bit of a mess).
session_context_ (in the enterprise build) is the SessionContextHandler.
session_ is the SessionHL which implements the bolt::Session interface.

SessionContextHandler actually depends on SessionInterface, an interface that bolt::Session inherits.

SessionContextHandler uses this Register to keep track of the active sessions and loop through them when deleting a database. 
SessionHL uses the SessoinContextHandler to do all the multi-tenency stuff.

Maybe we could remove the SessionInterface somehow. Got any suggestions?

Reply:
Currently no :) Need to dive into this more

Reply:
Actually the registration and looping through is a leftover from an older design. It still has a function but it may be easier to remove than I tough.
Everything is now contained in the InterpreterContext and the only thing OnDelete (the callback I loop trough) does is that is double checks that everything is fine.
I think if we change the synchronization a bit, we could have the sessions just increment/decrement a ref counter which we check in SessionContextHandler and delete if 0. 
There are still some problems with the defunct storage, but it could work.

I'll put a TODO here and resolve this comment.